src/lib.rs

```diff
@@ -836,50 +842,91 @@ impl<T> core::future::Future for Receiver<T> {
     type Output = Result<T, RecvError>;
 
     fn poll(self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> Poll<Self::Output> {
-        // FIXME: relax orderings and fix racy access to the waker if necessary
-
-        // SAFETY: The channel will not be freed while this method is still running.
+        // SAFETY: the existence of the `self` parameter serves as a certificate that the receiver
+        // is still alive, meaning that even if the sender was dropped then it would have observed
+        // the fact that we're still alive and left the responsibility of deallocating the
+        // channel to us, so `self.channel` is valid
         let channel = unsafe { self.channel_ptr.as_ref() };
 
-        match channel.state.load(SeqCst) {
+        // ORDERING: we use acquire ordering to synchronize with the store of the message.
+        match channel.state.load(Acquire) {
             // The sender is alive but has not sent anything yet.
-            EMPTY => unsafe { channel.write_async_waker(cx) },
+            EMPTY => {
+                // SAFETY: We can't be in the forbidden states, and no waker in the channel.
+                unsafe { channel.write_async_waker(cx) }
+            }
             // We were polled again while waiting for the sender. Replace the waker with the new one.
             RECEIVING => {
+                // ORDERING: We use relaxed ordering on both success and failure since we have not
+                // written anything above that must be released, and the individual match arms
+                // handle any additional synchronization.
                 match channel
                     .state
-                    .compare_exchange(RECEIVING, EMPTY, SeqCst, SeqCst)
+                    .compare_exchange(RECEIVING, EMPTY, Relaxed, Relaxed)
```

Comment:
I believe this is correct, but I just want to double-check with some people that the ordering doesn't need to be `Acquire, Relaxed` to prevent the dropping of the waker from being executed out of order. If there was no branch here, then this ordering would not be okay, and I'm not sure if the branch changes that.

Reply:
Would be great if you could get some more input on that. Because I'm not sure either. But it feels very weird if it can execute a drop before knowing that it's supposed to be dropped even. But modern CPUs are crazy, so why not :)

Reply:
Yeah I double checked and this ordering is fine.